lib/hls/hls_parser.js

```diff
@@ -1402,24 +1403,73 @@ shaka.hls.HlsParser = class {
     }
 
     let encrypted = false;
+    let clearKey;
     /** @type {!Array.<shaka.extern.DrmInfo>}*/
     const drmInfos = [];
     const keyIds = new Set();
 
-    // TODO: May still need changes to support key rotation.
+    /* eslint-disable no-await-in-loop */
+    // TODO: May still need changes to support key rotation and discontinuity.
```

Comment:
Yes, I see other players store key info per segment, so key rotation can be handled. I'm checking how to add key info to SegmentReference, but likely only for AES-128. DRM is handled outside of segment for a long time, so I don't know how much change is needed to support DRM key rotation, or how unencrypted discontinuity is currently handled in a DRM stream.

Reply:
Could you explain how unencrypted discontinuity is currently handled in a DRM stream? I could follow the same process.

Reply:
Discontinuity results in two side-by-side references having certain parameters change.  For example, different init segment references.

When StreamingEngine sees a change in init segment, timestamp offset, etc., it will handle that by adjusting offsets, fetching new init segments, etc.

So if a key can vary from segment to segment, you could put the key structure in SegmentReference instead of Stream.  Then StreamingEngine becomes responsible for fetching new keys, just like it would fetch new init segments.  (It is also smart enough not to fetch the same init segment over and over, so you should handle that for keys, too.)